### DISCUSSION TITLE
Conditional mutual information in pyagrum.

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#25] [Q&A] [answered]
Suppose I have a pandas dataframe that is a discrete dataset, how can I go about using the function `pyAgrum.InformationTheory.mutualInformationXYgivenZ()` WITHOUT having a Bayesian network? Any pointers or examples will be greatly appreciated!

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch ,
when dealing with data, a good idea is to look at `BNLearner` :

https://pyagrum.readthedocs.io/en/1.13.1/BNLearning.html#pyAgrum.BNLearner.mutualInformation

PS- oups, the description of "returns" is not correct ...

![image](https://github.com/agrumery/aGrUM/assets/5980636/bc6bd60a-025a-480e-9692-33c2348c5872)

**Reply by phwuil** [reply]
Fixed ! [c55b7f7](https://gitlab.com/agrumery/aGrUM/-/commit/c55b7f7cd024f640ee8cc710099ded31b7e16ea2)

**Comment by kenneth-lee-ch** [comment]
But how come it is not returning the actual mutual information instead of G
statistic (I am guessing its some sorts of CI tests)?

**Comment by phwuil** [comment]
It does return the actual (log2) mutual information. This is just the documentation that was wrong.

![image](https://github.com/agrumery/aGrUM/assets/5980636/7e0f8464-4d45-40a2-b7ce-708d81ff9868)

(G2 and chi2 both return the value of the statitistics and the p-value)

**Reply by kenneth-lee-ch** [reply]
@phwuil how can I fill the CPT just by frequency counts instead of the learning algorithms in BNLearner? How can I do that from pyagrum? or is BNLearner counting frequency? How does BNLearner derive mutual information? Any assumptions used along the process?

**Reply by phwuil** [reply]
Hi @kenneth-lee-ch , `fitParameters`allow this. If you want the "real" frequency count, you just use it directly. You can access to pseudo-count by adding a prior (like smoothing)

```python
bn=gum.fastBN("A->B<-C<-D",4)
gnb.flow.row(gnb.getInference(bn,size="3"),captions=["the original BN"])



size=10
print(f"Generate sample in dataframe of size {size}")
print("--------------------------------------------\n")

df,_=gum.generateSample(bn,size)
for n in "ABCD":
    print(f"P({n}) for pandas : {list(df.value_counts(n,sort=False)/len(df))}")
    
print() 
print("fit parameters with this dataframe")
print("----------------------------------\n")
bn2=gum.fastBN("A;B;C;D",4)

learner=gum.BNLearner(df,bn) # bn is used as a template for all the variables
learner.fitParameters(bn2)
gnb.flow.row(gnb.getInference(bn2,size="10"),captions=["bn2 after fitParameters"])


print() 
print("fit parameters with Prior (smoothing=1)")
print("----------------------------------\n")
learner=gum.BNLearner(df,bn) # bn is used as a template for all the variables
learner.useSmoothingPrior(1)
learner.fitParameters(bn2)
gnb.flow.row(gnb.getInference(bn2,size="10"),captions=["bn2 after fitParameters"])


print() 
print("fit parameters with Prior (smoothing=0.001)")
print("----------------------------------\n")
learner=gum.BNLearner(df,bn) # bn is used as a template for all the variables
learner.useSmoothingPrior(1e-3)
learner.fitParameters(bn2)
gnb.flow.row(gnb.getInference(bn2,size="10"),captions=["bn2 after fitParameters"])
```

![image](https://github.com/agrumery/aGrUM/assets/5980636/43ee3eb0-35bc-41d4-81ae-e2092393645c)

**Reply by kenneth-lee-ch** [reply]
@phwuil  Thank you for the response! Does that mean the Bayes net is part of the estimation? When it comes to computing the mutual information, does that also rely on the graph structure?

**Reply by phwuil** [reply]
When you `fitParameters(bn)`, of course the BN is part of the estimation. To have the marginals, I created a BN without arcs.

On the other hand, the statistics (G2, Chi2, MutualInformation, Entropy, etc.) directly address variables (columns) in the database. These statistics are part of the components that allows to build new structural learning algorithm on top of pyAgrum.

For instance : what is this algo ?

```python
def mysteriouAlgo(filename,alpha=0.1):  
    disc=skbn.BNDiscretizer()
    template=disc.discretizedBN(filename)
    learner=gum.BNLearner(filename,template)
  
    g=gum.UndiGraph.completeGraph(learner.nbCols())
    sepset=dict()
    for k in range(learner.nbCols()-2):
        for (x,y) in g.edges():
            neix=set(g.neighbours(x))
            neix.discard(y)
            if len(neix)>=k:
                for pss in itertools.combinations(neix,k):
                    _,pvalue=learner.G2(learner.nameFromId(x),
                                        learner.nameFromId(y),
                                        [learner.nameFromId(z) for z in pss])
                    
                    if pvalue>alpha:
                        g.eraseEdge(x,y)
                        sepset[(x,y)]=pss
                        break
        
    mg=gum.MixedGraph()
    mg.addNodes(learner.nbCols())
    for mid in g.nodes():
        for x,y in itertools.combinations(g.neighbours(mid),2):
            if not g.existsEdge(x,y):
                if mid not in sepset[x,y]:
                    mg.addArc(x,mid)
                    mg.addArc(y,mid)
    for x,y in g.edges():
        if not mg.existsArc(x,y) and not mg.existsArc(y,x):
            mg.addEdge(x,y)
          
    mr=gum.MeekRules()
    dag=mr.propagateToDAG(mg)
  
    bn=gum.BayesNet(template)
    for x,y in dag.arcs():
        bn.addArc(x,y)
    learner.fitParameters(bn)
    return bn
```    

this is a simple PC in python on top of pyAgrum (BNLearner and MeekRules) and skbn (for the automatic discretization of continuous variables)

**Reply by phwuil** [reply]
(by the way, all the computation of statistics (based on countings) are parallelized with pyAgrum)

Which allows this algo to be quite fast with large database (here 10 000 lines):

![image](https://github.com/agrumery/aGrUM/assets/5980636/3703386f-5100-43a1-92b0-7a57d22fead7)


And the result :

![image](https://github.com/agrumery/aGrUM/assets/5980636/4d24c3d9-c825-4d53-8724-6ba14191af98)

**Reply by kenneth-lee-ch** [reply]
@phwuil is there a way I can directly compute mutual information from the bn.cpt without samples?  How can I use these functions? https://pyagrum.readthedocs.io/en/1.9.0/BNToolsInformationTheory.html

**Reply by phwuil** [reply]
Hi @kenneth-lee-ch , here also, the doc must be improved ... :-(

![image](https://github.com/agrumery/aGrUM/assets/5980636/e874ec92-9f27-455d-9d1f-9877d4e33660)

**Reply by phwuil** [reply]
(in the next release)

![image](https://github.com/agrumery/aGrUM/assets/5980636/9cb91d93-9452-42d6-b36c-8128eb0c78e3)

**Reply by kenneth-lee-ch** [reply]
@phwuil thanks for all the help. I wonder if you know of any scalable way of computing conditional mutual information? The kind that can scale with 1000+ variables if exists..? When I use bn.learner, it basically kills the program when I condition on around 30 variables.

**Reply by phwuil** [reply]
Hi @kenneth-lee-ch, you have to take into account the inherent complexity of this computation :

![image](https://github.com/agrumery/aGrUM/assets/5980636/e895dba0-97ca-41cd-a17a-f52a6ed256ed)

IF Z consists in 30 variables, the sum over Z represents (at least) 2^30 iterations. Do not try this at home if Z contains 1000 variables :-)

The way pyAgrum organizes the computation of the conditional mutual information is to first compute p(x,y,z) in a unique inference because all the probabilities in the above formula are somehow marginalizations of this joint probability. Of course if Z contains >30 variables, this is not a correct solution since you need at least 2^30 double numbers in memory to store this joint probability.

With an inference in each iteration (rather than an inference before iterations), you could get rid of this memory problem but at the cost of a significant increase in computing time of each iteration ... And you still have at least 2^30 iterations to perform ...

**Comment by kenneth-lee-ch** [comment]
@phwuil I tried to directly derive mutual information based on a randomly generated bn via the following code below

```
import pyAgrum as gum
import numpy as np
bn = gum.randomBN(n=10) # 
ie = gum.LazyPropagation(bn)
names = list(bn.names())
n = len(names)
mi_matrix = np.zeros((n, n))

for i in range(n):
    for j in range(i+1, n):
        it = gum.InformationTheory(ie, names[i], names[j])
        mi = it.mutualInformationXY()
        mi_matrix[i, j] = mi

print("Pairwise Mutual Information Matrix:")
print(mi_matrix)
```

But I run into this error, shouldn't I use `it.mutualInformationXY()` for computing mutual information given any pair?
```
UndefinedElement: [pyAgrum] Undefined element: The variable X03(2) does not belong to this optimized inference.
```

**Comment by phwuil** [comment, marked as answer]
Hi, you found a problem : it seems that (a part of) the state of the inference is kept from an `InformationTheory` instance to another. For now, a (not so expensive) workaround is to create the `LazyPropagation` object in the loop :

```python
import pyAgrum as gum
import numpy as np

bn = gum.randomBN(n=10) # 
mi_matrix = np.zeros((bn.size(), bn.size()))

for i,ni in bn:
    for j,nj in bn:
        if i<=j:
            continue
        mi_matrix[i, j] =  gum.InformationTheory(gum.LazyPropagation(bn), 
                                                 ni,nj).mutualInformationXY()

print("Pairwise Mutual Information Matrix:")
print(mi_matrix)
```
(Note the loop in bn that iterates over all the pairs `(node_id,node_name)`.

**Reply by kenneth-lee-ch** [reply]
I see that some entries are negative values in the matrix. Is this an expected behavior?

**Reply by phwuil** [reply]
This is just floating point errors :

![image](https://github.com/user-attachments/assets/685bfddb-44ae-47aa-8f64-b06ca5c7284b)

**Comment by phwuil** [comment]
Note also that this code has a small problem : it assumes that the nodeids are integers from 1 to n. This is not always the case :  

![image](https://github.com/user-attachments/assets/a46c478e-e54f-4327-bbaf-312eb24af45b)

However, if you do not use the method `bn.erase(n)`, this assumption is valid.
